### PR TITLE
Implement support for SLUB Dresden

### PR DESCRIPTION
**Pull request by StefRe**
This API is specific for the new OPAC of Sächsische Landes- und Universitätsbibliothek in Dresden: www.slub-dresden.de.
See also #546.

### CODE REVIEW

**Comment by raphaelm** [comment]
I'm not sure what you did, but git can be nasty at times ;) I pulled your branch, re-based it again properly (git rebase -i master, then resolved two pseudeo-conflicts) and force-pushed it back again. You can force-pull it, if you want (but if it's done, you don't need to), but before you do anything to your local history, please double-check I included all your changes ;)

**Comment by StefRe** [comment]
After a long pause I finally sorted it out to correctly rebase and update the PR without having all old history appended here (thanks to @raphaelm  and [this](https://stackoverflow.com/a/45286781/3944322) SO answer).

Now I see the all new additions fail at CI with a long list of lint issues at Google Play Services Debug. Is this something I should worry about? As far as I can see none of them pertain to `libopac` code but it's a bit difficult to browse the 6000+ lines of `lint` output.

**Comment by johan12345** [comment]
Great, thank you!

The error in the Travis CI log seems to be:
```
de.geeksfactory.opacclient.apis.SLUBAccountTest > testParseEmptyAccountData FAILED
    org.junit.ComparisonFailure at SLUBTest.kt:23
```
so there is a failure in the SLUBTest. Maybe the parsing of `validUntil` is not working correctly?

**Comment by StefRe** [comment]
Oops, I completely overlooked this one in the sheer stream of lint warnings. The reason seems to be `LocalDateTime(it)` where `it `is a string "2020-03-31". Maybe it's a localization issue as the tests passes just fine on my German locale. I'll have to look into it.
CI says in line 1661 `There were failing tests. See the report at: file:///home/travis/build/opacapp/opacclient/opacclient/libopac/build/reports/tests/test/index.html`. Is there any way to retrieve this file to see what was the actual value or the detailed error?

**Comment by johan12345** [comment]
>  The reason seems to be LocalDateTime(it) where it is a string "2020-03-31"

Hmm, maybe `LocalDate(it)` would be sufficient if the time is not included anyway?

> Is there any way to retrieve this file to see what was the actual value or the detailed error?

Well, you can let Travis CI output that file by adjusting the `cat` command in [travis.yml](https://github.com/opacapp/opacclient/blob/master/.travis.yml).

**Comment by StefRe** [comment]
Thank you @johan12345, I solved this silly error of mine: I hardcoded the expected formatted date value in the test according to German style but it should have been formatted according to the current locale. So the code was OK, just the test was faulty.

**Comment by johan12345** [comment]
I quickly tested your API today and it already looks really good!

A few tips regarding the remaining `TODO` points:

- > //TODO: get status (one request per item!)

   Probably the best way to do this without affecting the performance too much is to use the `asyncGet` function implemented in the `OkHttpBaseApi`. Then, you can put all the `CompletableFuture`s you constructed into a list before calling `get()` on them, so that the requests are executed in parallel.

- `getSupportFlags`: Probably, you can at least support `OpacApi.SUPPORT_FLAG_ENDLESS_SCROLLING`. This just means that `getResultById` is not affected by which page of the result list was last loaded - so if you don't return it, the app will first go back to the corresponding page before loading the details.

- `filterResults`: This is not needed in the Open Source version of the app, you can just `return null`

- `reservation`, `cancel`: If these are supported by the OPAC, you can implement them, but we could also first merge without support for these, if you like.

- `getSupportedLanguages`, `setLanguage`: It seems that the SLUB OPAC supports German and English, so support for setting the language correctly would be nice (but not absolutely required).

**Comment by StefRe** [comment]
Thanks for taking the time and testing the API.

- `get status`: thanks for the suggestion, I'll certainly have a look a it and get back to you if I'm stuck here, but right now I guess this is of lower priority. This is due to the limited usefulness of the status as in many many cases it just says "click for more information about availability" (in the online web catalogue), so you'll have to open the details page anyway.

- `getSupportFlags`: thanks for the explanation, i didn't understand the implications of this flag. Will add `OpacApi.SUPPORT_FLAG_ENDLESS_SCROLLING`

- `filterResults`: `setFilters` seems easy to implement here but my understanding is that even if I implement `filterResults` it won't work in the FOSS version, right? If it will work I'll definetly try to implement it as due to the usually huge amount of results the search is not very useful without the possibility to further narrow it down to at least physical vs electronic copies.

- `cancel`: is implemented but not yet pushed: when I cancel an item it first does nothing (no dialog appears) and updates the data (which remain unchanged). Only when I cancel the item the second time the dialog "do you really want to cancel yes/no" appears and now it works correctly. Not sure if it's an issue of my `cancel `implementation or the app, the problem is I don't want to play around with this function too much on my account for obvious reasons :).

 - `reservation` is not yet implemented. With SLUB reservation is per copy, not per item and it's a two step process: you'll have to select the pick-up location from a list of offered locations. I haven't yet looked into it if this is supported by the app.
 
- `getSupportedLanguages`, `setLanguage`: it's true that the web catalogue offers English and German but the English version doesn't switch completely to English: all information concerning inter-library loans, most other account information and most of the captions/titles for links remain in German. So it mostly only translates the search and result field names. That's why this is of low priority for me but I'm going to add it later.

Although I used my debug version for over a year now I still find occasional errors and I have a couple of commits still sitting on my hard disk and also some uncommitted work and tests mainly regarding item details.  
For all kinds of reservations I'll have to study how the json information changes in the different stages of the process as it not always coincides with what is displayed in the account view of the online catalogue. I finished this for simple reservations of lent out items, for stack requests (Magazinbestellungen) and for request for usage on premises (Benutzung nur im Haus). Still working on the inter library loan process.   

One question about **cover images**: the catalogue doesn't supply thumbnails but in some cases there are links to cover images, mainly at http://swbplus.bsz-bw.de but also at the editors' pages. The size of these images typically ranges from about 40 kB to 1MB. Do you think I should include these URLs given that the images could be rather big in some cases? (for items from [Deutsche Fotothek](www.deutschefotothek.de) the API supplies thumbnails and I use them.)

I think I can finish the open issues by beginning of February so you can merge it and then I can complete any open items later. 

Lastly, from time to time I get leakage warnings (Dumping memory, app will freeze. Brrrr.) when I run the debug version on my phone. This is mainly after a new debug install, didn't get it later when running the app for a longer time. The strange thing is that I can't see further information in the Leaks app/icon on my phone anymore - it just says "Leaks in de.geeksfactory.opacclient.debug". This also happens if I don't query my account or search anything, so I guess this is more related to the app than to libopac. What couId I do here to narrow down the problem?

**Comment by johan12345** [comment]
> filterResults: setFilters seems easy to implement here but my understanding is that even if I implement filterResults it won't work in the FOSS version, right? If it will work I'll definetly try to implement it as due to the usually huge amount of results the search is not very useful without the possibility to further narrow it down to at least physical vs electronic copies.

Yeah, filterResults won't work in the FOSS app, the UI for that is only implemented in the Plus Edition. Most of the time, that is not such a big problem as fields such as media type are also available in the search form itself - but it seems that this is not the case in the SLUB OPAC. Maybe it would still be possible to add some of the filtering fields to the search form?

> One question about cover images: the catalogue doesn't supply thumbnails but in some cases there are links to cover images, mainly at http://swbplus.bsz-bw.de but also at the editors' pages. The size of these images typically ranges from about 40 kB to 1MB. Do you think I should include these URLs given that the images could be rather big in some cases? (for items from Deutsche Fotothek the API supplies thumbnails and I use them.)

Hm, do you have some links with examples for this? Sometimes there are even URL parameters to specify the size of the cover (like with the Amazon cover server, which some OPACs use), but this doesn't sound like it is the case here.
But I would say a 1MB cover is not a huge problem in the detail view (`getResultById`), I just wouldn't use it in the search result list.

> reservation is not yet implemented. With SLUB reservation is per copy, not per item and it's a two step process: you'll have to select the pick-up location from a list of offered locations. I haven't yet looked into it if this is supported by the app.

Yes, that's possible: the `ReservationResult` has the possibility to return some selections or confirmations that the user needs to do before the reservation is performed - you can see examples for this in most of the other APIs.

```java
ReservationResult res = new ReservationResult(MultiStepResult.Status.SELECTION_NEEDED);
res.setSelection(options);
res.setActionIdentifier(ReservationResult.ACTION_BRANCH);
return res
```

> cancel: is implemented but not yet pushed: when I cancel an item it first does nothing (no dialog appears) and updates the data (which remain unchanged). Only when I cancel the item the second time the dialog "do you really want to cancel yes/no" appears and now it works correctly. Not sure if it's an issue of my cancel implementation or the app, the problem is I don't want to play around with this function too much on my account for obvious reasons :).

Hmm, interesting... If you push this change somewhere I can take a look, it doesn't sound like this is a general bug in the app.

> Lastly, from time to time I get leakage warnings (Dumping memory, app will freeze. Brrrr.) when I run the debug version on my phone. This is mainly after a new debug install, didn't get it later when running the app for a longer time. The strange thing is that I can't see further information in the Leaks app/icon on my phone anymore - it just says "Leaks in de.geeksfactory.opacclient.debug". This also happens if I don't query my account or search anything, so I guess this is more related to the app than to libopac. What couId I do here to narrow down the problem?

That's probably not due to your changes - we added [LeakCanary](https://square.github.io/leakcanary/) to the debug version of the app [a long time ago](https://github.com/opacapp/opacclient/commit/da9fa72eaba508278d3d17797b248932ae6d82a9) but never had much time to further investigate the leaks that it is sometimes showing (and it wasn't a high priority as we never heard of any real issues arising from these). Sorry!

**Comment by StefRe** [comment]
One question about strings in libopac: most of the strings from [` interface StringProvider`](https://github.com/opacapp/opacclient/blob/master/opacclient/libopac/src/main/java/de/geeksfactory/opacclient/i18n/StringProvider.java) are implemented in [`strings_api_errors.xml`](https://github.com/opacapp/opacclient/blob/master/opacclient/opacapp/src/main/res/values/strings_api_errors.xml), including many strings that aren't errors but just information. Only three strings (`status`, `free_search `and `no_results`) come from [`strings.xml`](https://github.com/opacapp/opacclient/blob/master/opacclient/opacapp/src/main/res/values/strings.xml). Where should I put my implementation of general information strings (e.g. "ready for pick-up since ...", i.e. mostly status related information) - in `strings.xml` (where I thought it should be) or in `strings_api_errors.xml` (where de facto almost all strings are placed, be it error related or not)?

**Comment by johan12345** [comment]
`strings_api_errors.xml` is (despite the name) made for everything that comes from `StringProvider` - where originally most of the strings were error messages. The reason why we are separating these is that Android Lint would mark these as unused as we are not referencing their IDs in the `AndroidStringProvider` class.

The three strings that are in `strings.xml` despite this are probably there because they were already used by something else in the app.

**Comment by StefRe** [comment]
> Hm, do you have some links with examples for this?

example of cover image: https://katalog.slub-dresden.de/id/0-81817546X/ with a [780 kB image](http://www.netread.com/jcusers2/bk1388/319/9781472533319/image/lgcover.9781472533319.jpg)

> Hmm, interesting... If you push this change somewhere I can take a look,

I pushed it to this PR as dde1bba.

**Comment by StefRe** [comment]
I tested my `cancel `implementation dde1bba in a command line application like this:

        OpacApi.CancelResult cr = api.cancel(id, account, 0, null);

It  worked correctly and returned the following cancel result:
```
{
  "status" : "OK",
  "selection" : null,
  "details" : null,
  "message" : null,
  "actionIdentifier" : 0
}
```
For the app I'd prefer however to be asked if I really want to cancel the reservation. Do I have to process the `useraction `and `selection `parameters in some way for this? I looked in some of the other API implementations but couldn't find anything there.

**Comment by johan12345** [comment]
> cancel: is implemented but not yet pushed: when I cancel an item it first does nothing (no dialog appears) and updates the data (which remain unchanged). Only when I cancel the item the second time the dialog "do you really want to cancel yes/no" appears and now it works correctly. Not sure if it's an issue of my cancel implementation or the app, the problem is I don't want to play around with this function too much on my account for obvious reasons :).

Hm, I looked at your `cancel` implementation and found no obvious mistake. The only problem I could think of may have been that the account data hasn't been refreshed recently when you tested it. We require a maximum age of the account data (2 hours I think) to make sure that the item you want to cancel is actually still there.
https://github.com/opacapp/opacclient/blob/9872e4c8adf26d24083343eb5f66463cbf0fb53e/opacclient/opacapp/src/main/java/de/geeksfactory/opacclient/frontend/AccountFragment.java#L572

The dialog "do you really want to cancel" should always be shown by the app before actually calling your API's `cancel` implementation. You can show additional dialogs by returning status codes like `CONFIRMATION_NEEDED` and then checking the `useraction` and `selection` parameters to see that they were confirmed - but in this case that shouldn't be necessary. That's more commonly used for showing things like warnings that you have to pay a fee when you do a reservation (or, in some libraries, also prolongation).

**Comment by johan12345** [comment]
> example of cover image: https://katalog.slub-dresden.de/id/0-81817546X/ with a 780 kB image

I think that size is fine for showing it on the detail page. (just not in the result list - but the covers are only linked on the detail page anyway)

**Comment by StefRe** [comment]
> Hm, I looked at your cancel implementation and found no obvious mistake. 

Everything is OK, my `cancel` implementation actually works perfectly fine when run with no debugger attached. I first run the program under android studio's debugger (on a real device) which obviously interferes with the program in some strange way to produce these side effects. @johan12345: Sorry for bothering you with that and thanks for your time and effort you spent looking into it.

**Comment by StefRe** [comment]
@johan12345  - regarding this last commit b5ea66f:   
The `copy` string was introduced in 213fff62e. Now I get a warning `Overriding '@string/copy' which is marked as private in androidx.preference:preference.`. As everything worked fine so far I followed Android studios's recommendation `If deliberate, use tools:override="true", otherwise pick a different name.` without fully understanding what I'm doing here. Hope this is OK.

**Comment by johan12345** [comment]
Hm, that sounds like the string `copy` (probably in the sense of the verb, i.e. `copy = kopieren`) is already defined in the preference support library. I'm not sure where that string is actually used by that library and whether that appears in our app, but just to be safe I would suggest to pick a different name instead of overriding it. Thanks!

**Comment by StefRe** [comment]
@johan12345 done, see 3b7ca5e.

**Comment by StefRe** [comment]
I've been using with my debug version for over a year now without any problems, so I think you can review this PR for merging. All methods except for `setLanguage ` and `filterResults` are implemented and working. Test coverage is 86 % of LOC; if need be I can write more tests to increase coverage.   

There are still some TODOs, but I don't feel like being able to close them in the next couple of weeks. I may get back to them later.
* line # 155: get status: low priority due to the limited usefulness of the status as in many many cases it just says "click for more information about availability" (in the online web catalogue), so you'll have to open the details page anyway.

* line # 160: filterResults: not to implement in FOSS version

* line # 251: references: libero links must be resolved to `rsn `entries in current SLUB catalog and they then in turn to `id `entries, similar to next TODO line # 431

* line # 431/487: `id `of reserved/lent items: `bc `entries to be resolved to `id `entries by intercepting redirection

* line # 433: actual ready date of stack request: works but could be refined, need more examples to examine

* line # 479: permanent loans: need example, can't work on this as permantent loans are not granted to external users

* line # 489: inter-library loans renewability: need example (`X_is_flrenewable` field exists but json but not sure if it's actually used/properly set)

* line # 545/568: getSupportedLanguages/setLanguage: low priority as all information concerning inter-library loans, most other account information and most of the captions/titles for links remain in German even in the English version of the online catalog

**Review by johan12345 — CHANGES_REQUESTED** [review]
Hi @StefRe! Thank you for your hard work on the SLUB API and sorry for taking so long for the review. Your implementation looks very good to me and it is also great that you added good tests - which is especially helpful as we don't have an account in the library to be able to test account features.

I have added a few minor comments below, most of which can probably be addressed very easily. I haven't compiled and tested it on a device yet, but will of course do so before merging.

I agree that the remaining TODOs are all not essential - for some of them I added comments below with suggestions on how they could be implemented (but that doesn't mean that it needs to be done before we merge this).

Could you also please rebase your branch against the current master? GitHub says there are some conflicts (shouldn't be anything too complicated, probably just a few additions to the strings XML files that we have done in the meantime).

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt:89]: I'm afraid the clear texts for the pickup point codes will have to be hardcoded here. The API only sends the codes (without the text). The web catalog does of course show the texts when you're going to reserve an item or put a hold on it but it only shows the points available for this particular item (as a rule, just 1 or 2 possible pickup points). So I had to compile this list by trying to reserve a lot of different items and then see what code - clear texts combinations the web site offered.  
New pickup points, which are not in the map, will get their code as clear text:
https://github.com/opacapp/opacclient/blob/08dfa091c944603cda0a76cd025fb2c200d7ba14/opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt#L334
When this happens this new code must be looked up in the web catalog by trying to make the reservation and then update the code here.  
I understand that this is not ideal but it's the price for using the super fast API rather than scraping the relatively slow web site.

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt:126]: I changed it on purpose from `addQueryParameter` to `addEncodedQueryParameter` in 0d52c8b5bb77b because of issues when the query string contains spaces (see second part of last item in commit message). Umlauts and &, /, # etc work OK.

File: opacclient/libopac/src/test/java/de/geeksfactory/opacclient/apis/SLUBTest.kt
Comment (StefRe) [reply opacclient/libopac/src/test/java/de/geeksfactory/opacclient/apis/SLUBTest.kt]: Oh, yes - this is perfectly enough for testing, I kind of overcomplicated things :).  
Fixed in eedeeba.

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt:316]: Same comment as [above](https://github.com/opacapp/opacclient/pull/553#discussion_r502967615) applies.

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt]: Yes, fixed in a18564d (I added a comment in the source code)

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt:431]: This is an excellent idea 👍!  I was already a bit unhappy with having to resolve the barcode (bc) to id for every lent or reserved item in every account request just for the relatively unlikely case that it's really going to be used, i.e. that the user wants to look up the catalog entry for an item he has already lent or reserved.  
So I'd need to add a prefix to the id field (`id`, `bc` or maybe `rsn`) and then, if this prefix is not `id`, get the `id` from `bc` in `getResultById` by intercepting the redirection. In `getShareUrl` (I guess you meant this by `getShareLink`) I could simply return a `bc` or `rsn` link instead of an `id` link.  
I'm not sure how fast I can do this, so maybe we can leave this for later, but I'm going to implement it this way.

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt:487]: same as [above](https://github.com/opacapp/opacclient/pull/553#discussion_r503116266). Thanks for the suggestion.

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt:155]: Thanks for the suggestion, this looks very interesting although I'll have to read up a bit on asynch requests and completeable futures first.  I made a comment in the open items list [above](https://github.com/opacapp/opacclient/pull/553#issuecomment-701936760).

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt]: Yes, thanks -  `.forEach` is indeed better here, fixed in c81415a.

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt:141]: Yes, you're right for this particular case (`docs`), I should have returned an empty list here if the `docs` entry is missing. It **seems** that the jsons I get from the API are generally well formed, i.e. contain all the keys even if they have no value, but I'm not completely sure about it as the API sometimes returns some strange or even malformed jsons (seen so far mostly for journal articles or "grey" literature as theses etc.). This is why I decided to "be on the safe side" and generally use the `opt...` variants of the json functions. But you're right - I'll have to re-check if null values are OK in all cases. So I'll review each case and get back to you here later.

**Comment by StefRe** [comment]
@johan12345 Thanks for the detailed comments in the review, I resolved all but one where I need some more time. Regarding rebasing: I rebased it already on Sep 26 and it's no problem the rebase it again onto the current master but I think I read somewhere that you should never rebase a branch under review as it could mess up the review history/comments (not sure if this is still correct for current versions of git), so I'd like to wait with rebasing until everything is resolved and you're prepared to merge.

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt:141]: regarding the iterator: as far as I understand you'd need it to iterate over the JSONArray in a for loop, as in `for(doc in json.getJSONArray("docs")` but in this case I'd have to first create an empty list of SearchResults, then one by one append each SearchResult in the for loop and then return this list. Maybe I'm wrong but mapping SearchResult objects to a list seems more idomatic to me. I tried to use some extension functions from https://github.com/android/android-ktx/issues/241 but couldn't get it working. So I guess I'll stick with my solution.

**Comment by StefRe** [comment]
@johan12345 I resolved all review comments above and will wait for your go-ahead before rebasing onto the current master, see [previous comment](https://github.com/opacapp/opacclient/pull/553#issuecomment-707009395).
Further I prepared a PR for the SLUB config file, see https://github.com/opacapp/opacapp-config-files/pull/13.

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (johan12345) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt:89]: Okay, in that case I think it's fine. Thanks for the explanation!

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (johan12345) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt:431]: Yes, that's what I meant :) Sure, we can leave this for later.

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (johan12345) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt]: I added some extension functions in ad751d41700fac47cf818ba336a34bb316bde488, which should work. You should now be able to use something like
```kotlin
json.getJSONArray("docs")
                .map<JSONObject, SearchResult> {
                    //...
                }
```

**Review by johan12345 — APPROVED** [review]
Thanks! I think after rebasing and implementing the extension functions (see comment at https://github.com/opacapp/opacclient/pull/553#discussion_r504457783) this should be good to go :)

**Comment by StefRe** [comment]
@johan12345 Your commit e573c2c56 breaks `VuFindSearchTest`:
https://github.com/opacapp/opacclient/blob/e9e14f280b516bd8c30f34fa95d4e141fb6840eb/opacclient/libopac/src/test/java/de/geeksfactory/opacclient/apis/VuFindSearchTest.java#L51-L52
(`parseDetail` has 5 instead of 4 arguments now)  

How did you manage to compile the project?

**Comment by johan12345** [comment]
oops... yes, indeed, I forgot to adjust the tests after that commit. Building the app itself worked though. Will fix that tonight...

File: opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt
Comment (StefRe) [reply opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt]: Thanks, this works great although I had to fix the index range in #594.

**Comment by johan12345** [comment]
Tests should be fixed now with 85a67906d0c0a841bc9b6008cdaee46930077204.

**Comment by StefRe** [comment]
> Could you also please rebase your branch against the current master? 

I'm having difficulties in rebasing. The last rebasing on Sep 26 went smoothly without any problems (I just had to resolve some conflicts in build.gradle and in the string xml files). If I try to rebase now, I not only need to resolve the same conflicts again (which is very strange) but also need to resolve conflicts for each of my 55 commits of SLUB.kt and SLUBTest.kt relative to the preceding versions of these files. This in principle is no problem but at some point (at commit 22 of 55) git starts to complain that it can't execute the todo command (next pick). I tried some recommendations from stackoverflow but not fully understanding what I'm doing I didn't succeed.

> GitHub says there are some conflicts

This is strange as github says "This branch has no conflicts with the base branch". I also forked the current master this afternoon and merged my feature branch in it without any problems (https://github.com/StefRe/opacclient/tree/merge_test, this also includes the use of the new extension functions).  

**So my question is whether you could merge this PR as is**. I could then make a new PR for commit c79243ba50df (use utility functions for JSONArrays) against the new master.

**Comment by johan12345** [comment]
Ah, interesting. I had selected "Rebase and merge", and there GitHub says there are conflicts:
image

But if I say "Squash and merge" or "create a merge commit", it works:
image

So I'll try.

**Comment by johan12345** [comment]
That worked, thanks again! 🎉 I also cherry-picked your commit to add the JSONArray extension functions (27395799f8318c476a72162f2f1ee8ba2c7325a5).

**Pull request merged**
